File: leetcode/cpp_solutions/container_with_most_water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int end = height.size() - 1;
        int start = 0;
        int max_area = 0;
        while (start < end){
            if (min(height[start], height[end]) * (end - start) > max_area){
                max_area = min(height[start], height[end]) * (end - start);
            }
            else if (height[start] >= height[end]){
                end--;
            }
            else if (height[end] > height[start]){
                start++;
            }
        }
        return max_area;
    }
};

File: leetcode/golang/container-with-most-water.go
func Min(x int, y int) int {
    if x > y { return y }
    return x
}

func calcArea(height []int, pt1 int, pt2 int) int {
    breath := Min(height[pt1], height[pt2])
    if pt1 > pt2 { return pt1 - pt2 * breath }
    return (pt2 - pt1) * breath
}

func maxArea(height []int) int {
    var pt1 int = 0
    var pt2 int = len(height) - 1
    var area int 

    var maxArea int 
    for pt1 != pt2 {
        if area = calcArea(height, pt1, pt2); area > maxArea {
            maxArea = area
        } else if height[pt1] > height[pt2] {
            pt2--
        } else {
            pt1++
        } 
    }
    return maxArea
}
